### PR TITLE
[WIP] Global pool models

### PR DESCRIPTION
**Pull request by danielmkm**
Don't merge this, would like to use this as a collaborative space for @johngrantuk @brunoguerios. 

I'd like us to strive for a global model to represent pools moving forward, a model that should be leveraged across the entire stack. The introduction of the API allows us to define a more flexible model than is available via the subgraph alone.

Key areas of focus:
- The data model should facilitates “stupid” UIs 
  - So much of the code we end up writing is in transforming data from one form to another. It is error-prone and increases complexity. If we can front load more of this work in the API, it will allow consuming applications to be significantly lighter (simpler).
- BasePool model with extension for specific pool types rather than a single pool model with a ton of different optional fields
  - B-SDK: https://github.com/balancer/b-sdk/blob/main/src/data/types.ts#L18-L87
  - Beets API: https://github.com/beethovenxfi/beethovenx-backend/blob/v3-main/modules/pool/pool.gql
- The pool model itself should tell you what tokens you can use for invest/withdraw, as opposed to having to do a traversal of nested tokens.
  - Additionally, the pool model should itself say what tokens can be invested/withdrawn proportionally in a model that is easy to parse and traverse.
- In what format do we pass `numbers` when passing data between layers of the tech stack (API / frontend / SDK)? We decide on a standard and stick to it.
- Composable pool support - Id propose we enforce a maximum nesting of 2 levels (Composable/Weighted → Composable → Linear), anything deeper we don’t support.
- Future proof - in scenarios where key architecture components change, how do we model in such a way that it can be accommodated.

### CODE REVIEW

**Comment by danielmkm** [comment]
The `joinConfig` is the focus currently. How do we model a config that makes it significantly simpler to manage generalized pool joins, both in the SDK and in the UI. Currently the SDK has to do a significant amount of heavy lifting because the data is returned from the subgraph in a suboptimal way. Given the flexibility of the API, what is a more optimal data structure that makes the SDK code 10x simpler? @johngrantuk @brunoguerios

**Review by brunoguerios — COMMENTED** [review]
I like the overall idea of trying to design the pool model in a way that facilitates how the SDK and the FE handle generalised joins and exits. 😊
Thanks for the initial draft! It does help to start moving things.
I'll add comments to what called my attention so far and I hope it helps us move forward.

File: src/schema/pool-schema.ts [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+  poolTokenConfigs: BalancerPoolTokenJoinConfig[];
+
+  // the ui doesn't use proportional joins, it uses EXACT_TOKENS_IN_FOR_BPT_OUT to allow the
+  // user to configure any amount for each pool token, this concept could maybe use a better name.
+  // Additionally, the ui provides a proportional estimation when applicable.
+  //
+  // Since it's possible not all tokens are being used for a given join, this path is not always the
+  // "most efficient" for a given combination of tokens, but for the sake of limiting the scope, we
+  // provide a "generic" path that works for any given combination.
+  steps: BalancerPoolJoinStep[];
```
[the remark this answers is not in the document]
Comment (danielmkm) [reply src/schema/pool-schema.ts:222]: An illustrative example with bbaUSD / WETH:

Steps:
batchSwap -> joinPool -> joinPool

In an instance where tokensIn: [bbaUSD, WETH], you would disregard both the batchSwap and first joinPool step, performing on the last `joinPool`

File: src/schema/pool-schema.ts [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  leafTokens: Address[];
+}
+
+export type BalancerPoolJoinStep =
+  | BalancerPoolJoinExitBatchSwapStep
+  | BalancerPoolJoinMintBptStep;
+
+export type BalancerPoolExitStep =
+  | BalancerPoolJoinExitBatchSwapStep
+  | BalancerPoolExitBurnBptStep;
```
[the remark this answers is not in the document]
Comment (danielmkm) [reply src/schema/pool-schema.ts:292]: True, there's a few possible ways to model this.

Implicitly you could say that if you're leaf token is inside of a linear pool, it is known there there is  a "possible" unwrap step.

Explicitly the batchSwap step would have some additional metadata to flag the tokens originating from a linear pool to act as an identifier that some additional logic needs to be done to identify if an unwrap is necessary.

File: src/schema/pool-schema.ts [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+  tokensOut: Address[];
+}
+
+// The mint BPT step is a standard join step. At least one mint step will always occur on the pool itself.
+// Additionally, a second mint step will be present if there is a nested stable BPT in the pool.
+export interface BalancerPoolJoinMintBptStep {
+  __typename: "BalancerPoolJoinMintBptStep";
+  poolId: ID;
+  tokensIn: Address[];
+}
```
[the remark this answers is not in the document]
Comment (danielmkm) [reply src/schema/pool-schema.ts:317]: I would say not, but just my opinion. I think the API is trying to provide a model to make your life easier, ie: give you data in a form that is easier to work with than currently provided by the subgraph.

**Pull request closed without merging**